#! /bin/sh /usr/share/dpatch/dpatch-run
## 02_processes_parsing.dpatch by  <darrell@swiftstack.com>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fix process name parsing.

@DPATCH@

diff --git a/src/processes.c b/src/processes.c
index 8f4eb88..c9a439e 100644
--- a/src/processes.c
+++ b/src/processes.c
@@ -880,10 +880,10 @@ int ps_read_process (int pid, procstat_t *ps, char *state)
 
 	char *fields[64];
 	char  fields_len;
+	char *bufp, *namep;
 
 	int   i;
 
-	int   name_len;
 
 	derive_t cpu_user_counter;
 	derive_t cpu_system_counter;
@@ -900,6 +900,21 @@ int ps_read_process (int pid, procstat_t *ps, char *state)
 		return (-1);
 	buffer[i] = 0;
 
+	/* Parse out ps->name while sanitizing any whitespace ( \t\r\n) in
+		the process name before calling strsplit
+	 */
+	for (bufp = buffer; *bufp != '\0' && *bufp != '('; bufp++);
+	if (*bufp == '\0' || *(++bufp) == '\0')
+		return (-1);
+	namep = ps->name;
+	while (*bufp != '\0' && *bufp != ')') {
+		*namep++ = *bufp;
+		if (*bufp == ' ' || *bufp == '\t' || *bufp == '\r' || *bufp == '\n')
+			*bufp = '_';
+		bufp++;
+	}
+	*namep = '\0';
+
 	fields_len = strsplit (buffer, fields, STATIC_ARRAY_SIZE (fields));
 	if (fields_len < 24)
 	{
@@ -909,18 +924,6 @@ int ps_read_process (int pid, procstat_t *ps, char *state)
 		return (-1);
 	}
 
-	/* copy the name, strip brackets in the process */
-	name_len = strlen (fields[1]) - 2;
-	if ((fields[1][0] != '(') || (fields[1][name_len + 1] != ')'))
-	{
-		DEBUG ("No brackets found in process name: `%s'", fields[1]);
-		return (-1);
-	}
-	fields[1] = fields[1] + 1;
-	fields[1][name_len] = '\0';
-	strncpy (ps->name, fields[1], PROCSTAT_NAME_LEN);
-
-
 	*state = fields[2][0];
 
 	if (*state == 'Z')
